File: backend/src/main/java/com/noisevisionsoftware/nutrilog/model/diet/template/DietTemplate.java
package com.noisevisionsoftware.nutrilog.model.diet.template;

import com.google.cloud.Timestamp;
import lombok.*;

import java.util.List;
import java.util.Map;

@Data
@Builder
@AllArgsConstructor
@NoArgsConstructor
public class DietTemplate {
    private String id;
    private String name;
    private String description;
    private DietTemplateCategory category;
    private String createdBy;
    private Timestamp createdAt;
    private Timestamp updatedAt;
    private int version;

    private int duration;
    private int mealsPerDay;
    private Map<String, String> mealTimes;
    private List<String> mealTypes;

    private List<DietTemplateDayData> days;
    private DietTemplateNutrition targetNutrition;

    private int usageCount;
    private Timestamp lastUsed;
    private String notes;
}

File: frontend/src/types/DietTemplate.ts
export interface DietTemplate {
    id: string;
    name: string;
    description?: string;
    category: DietTemplateCategory;
    categoryLabel: string;
    createdBy: string;
    createdAt: string;
    updatedAt: string;
    version: number;
    duration: number;
    mealsPerDay: number;
    mealTimes: Record<string, string>;
    mealTypes: string[];
    days: DietTemplateDay[];
    targetNutrition?: DietTemplateNutrition;
    usageCount: number;
    lastUsed?: string;
    notes?: string;
    totalMeals: number;
    totalIngredients: number;
    hasPhotos: boolean;
}

export interface DietTemplateDay {
    dayNumber: number;
    dayName: string;
    notes?: string;
    meals: DietTemplateMeal[];
}

export interface DietTemplateMeal {
    name: string;
    mealType: string;
    time: string;
    instructions?: string;
    ingredients: DietTemplateIngredient[];
    nutritionalValues?: any;
    photos: string[];
    mealTemplateId?: string;
}

export interface DietTemplateIngredient {
    name: string;
    quantity: number;
    unit: string;
    original: string;
    categoryId?: string;
    hasCustomUnit: boolean;
}

export interface DietTemplateNutrition {
    targetCalories?: number;
    targetProtein?: number;
    targetFat?: number;
    targetCarbs?: number;
    calculationMethod?: string;
}

export enum DietTemplateCategory {
    WEIGHT_LOSS = 'WEIGHT_LOSS',
    WEIGHT_GAIN = 'WEIGHT_GAIN',
    MAINTENANCE = 'MAINTENANCE',
    SPORT = 'SPORT',
    MEDICAL = 'MEDICAL',
    VEGETARIAN = 'VEGETARIAN',
    VEGAN = 'VEGAN',
    CUSTOM = 'CUSTOM'
}

export interface CreateDietTemplateRequest {
    name: string;
    description?: string;
    category: string;
    duration: number;
    mealsPerDay: number;
    mealTimes: Record<string, string>;
    mealTypes: string[];
    days?: DietTemplateDay[];
    targetNutrition?: DietTemplateNutrition;
    notes?: string;
    dietData?: any;
}

export interface DietTemplateStats {
    totalTemplates: number;
    templatesByCategory: Record<string, number>;
    mostUsedTemplate?: DietTemplate;
    newestTemplate?: DietTemplate;
    totalUsageCount: number;
}